scanpipe/pipes/alpine.py

```diff
@@ -20,7 +20,91 @@
 # ScanCode.io is a free software code scanning tool from nexB Inc. and others.
 # Visit https://github.com/nexB/scancode.io for support and download.
 
+import json
+
 from packagedcode import alpine
+from fetchcode.vcs.git import fetch_via_git
+from fetchcode import fetch
+from shutil import copytree
+from scanpipe.models import DiscoveredPackage
+
+APORTS_URL = "https://gitlab.alpinelinux.org/alpine/aports.git"
+APORTS_DIR_NAME = "aports"
+APORTS_SUBDIRS = ["main", "non-free", "community", "testing", "unmaintained"]
+
+
+def download_or_checkout_aports(aports_dir_path, alpine_version, commit_id=None):
+    """
+    Download aports repository and it's branch based on `alpine_version`.
+    Checkout to a branch (alpine version).
+    If `commit_id` is provided also checkout to a commit.
+    Return False if checkout failed otherwise True. #TODO Proper fetchcode patch required (extending #54)
+    """
+    ver = alpine_version.split(".")
+    aports_dir_path = str(aports_dir_path / APORTS_DIR_NAME)
+    fetch_via_git(f"git+{APORTS_URL}@{ver[0]}.{ver[1]}-stable", aports_dir_path)
+    if commit_id:
+        fetch_via_git(f"git+{APORTS_URL}@{commit_id}", aports_dir_path)
+    return True
+
+
+def get_packages_from_db(project):
+    """
+    Get alpine packages from the database that belong to a `project`
+    Yield only root and not yet scanned packages along with paths needed when performing a scan.
+    """
+    for package in DiscoveredPackage.objects.project(project):
+        scan_id = f"{package.name}_{package.version}"
+        scan_target_path = project.tmp_path / scan_id
+        scan_result_path = project.output_path / (scan_id + ".json")
+        if (
+            package.type == "alpine"
+            or (package.source_packages and package.source_packages[0] in package.purl)
+            or not scan_result_path.exists()
+        ):
+            yield scan_target_path, scan_result_path, package
+
+
+def prepare_scan_dir(package_name, scan_target_path, aports_dir_path=None):
+    """
+    Find package's aports path and if found execute the following steps:
+        Copy all the files from that path into `scan_target_path`
+        Download all package's sources into `scan_target_path`
+    The default value of `aports_dir_path` is set to the parent of the `scan_target_path`
+    """
+    if aports_dir_path is None:
+        aports_dir_path = scan_target_path.parent
+    for subdir_name in APORTS_SUBDIRS:
```

Comment:
Rather than trying all the paths and since each package name should be unique within a given release what about a different logic:
1. walk the aports tree once and build a mapping of `{<package name>: <package path> }` (or maybe even better <APKBUILD path>` )
2. use this in your loop below

Reply:
I think current solution might be quicker (as we're not iterating over all packages in alpine - just checking for existance of pkgs we're interested in. the APORTS_SUBDIRS is sorted by probability of having package we're looking for (main->non-free->contrib...)

Reply:
So adding more explanations in doc strings will do?